addons/website/models/website.py

```diff
@@ -1340,3 +1341,235 @@ def _get_cached_values(self):
+                for record in records:
+                    for field, value in record.items():
+                        if isinstance(value, str):
+                            value = value.lower()
+                            if field == 'arch_db':
+                                value = self._search_text_from_html(value)
+                            for word in re.findall(match_pattern, value):
+                                yield word
+
+        def basic_enumerate_words():
+            """
+            Browses through all words that need to be compared to the search term.
+            It extracts all words of every field associated to models in the fields_per_model parameter.
+            """
+            match_pattern = '\\w{%s,}' % min(4, len(search) - 3)
+            first = escape_psql(search[0])
+            for search_detail in search_details:
+                model_name, fields = search_detail['model'], search_detail['search_fields']
+                model = self.env[model_name]
+                if search_detail.get('requires_sudo'):
+                    model = model.sudo()
+                domain = search_detail['base_domain'].copy()
+                fields_domain = []
+                fields = ['arch_db' if field == 'arch' else field for field in fields if field in model]
+                for field in fields:
+                    fields_domain.append([(field, '=ilike', '%s%%' % first)])
+                    fields_domain.append([(field, '=ilike', '%% %s%%' % first)])
+                    fields_domain.append([(field, '=ilike', '%%>%s%%' % first)]) # HTML
+                domain.append(OR(fields_domain))
+                domain = AND(domain)
+                records = model.search_read(domain, fields, limit=1000)
+                for record in records:
+                    for field, value in record.items():
+                        if isinstance(value, str):
+                            value = value.lower()
+                            if field == 'arch_db':
+                                value = self._search_text_from_html(value)
+                            for word in re.findall(match_pattern, value):
+                                if word[0] == search[0]:
+                                    yield word.lower()
+
+        if len(search) < 4 or ' ' in search:
+            return search
+        search = search.lower()
+        words = {}
+        best_score = 0
+        best_word = False
+        for word in word_list or (trigram_enumerate_words() if self.env.registry.has_trigram else basic_enumerate_words()):
+            if search in word:
+                return search
+            if word[0] == search[0] and word not in words:
+                similarity = similarity_score(search, word)
+                if similarity > best_score:
+                    best_score = similarity
+                    best_word = word
+                words[word] = similarity
```

Comment:
Why store the `similarity` when it's unused? Why not make `words` a set?

Or if we're storing the similarity then surely we should just shove all the information in the dict and get the highest score afterwards?